File: euler_step_counter.rb
class EulerStepCounter
  TRANSITION_TOLERANCE = 0.6

  attr_reader :current_angle, :previous_angle, :diff, :rotations

  def initialize
    @current_angle = @previous_angle = @diff = @rotations = 0.0
  end

  # Update with the latest euler angle reading
  def update(angle)
    @previous_angle = current_angle
    @current_angle = angle

    if completed_a_revolution?
      @rotations += 1
    end

    @diff = current_diff
  end

  # The current position including rotations
  def linear_position
    current_angle + rotations
  end

private

  def completed_a_revolution?
    transition? && change_exceeds_threshold?
  end

  def current_diff
    current_angle - previous_angle
  end

  def absolute_diff
    if current_angle > previous_angle
      current_angle - previous_angle
    else
      previous_angle - current_angle
    end
  end

  def sign_of(number)
    number <=> 0
  end

  def transition?
    sign_of(current_diff) != sign_of(diff)
  end

  def change_exceeds_threshold?
    absolute_diff >= TRANSITION_TOLERANCE
  end
end

File: EulerStepCounter.cpp
#include "EulerStepCounter.h"

#define EULER_TRANSITION_TOLERANCE 0.6

EulerStepCounter::EulerStepCounter() :_previous_angle(0.0), _current_angle(0.0), _diff(0.0), _rotations(0.0) {}

void EulerStepCounter::update(float angle) {
  _previous_angle = _current_angle;
  _current_angle = angle;

  if( completed_a_revolution() == true) {
    _rotations = _rotations + 1.0;
  }

  _diff = current_diff();
}

float EulerStepCounter::linear_position() {
  return(_current_angle + _rotations);
}

float EulerStepCounter::current_diff() {
  return(_current_angle - _previous_angle);
}

float EulerStepCounter::absolute_diff() {
  if(_current_angle > _previous_angle)
    return(current_diff());
  else
    return(_previous_angle - _current_angle);
}

bool EulerStepCounter::completed_a_revolution() {
  return(transitioned() == true && change_exceeds_threshold() == true);
}

bool EulerStepCounter::transitioned() {
  return(current_diff() < 0.0 && _diff > 0.0 || current_diff() > 0.0 && _diff < 0.0);
}

bool EulerStepCounter::change_exceeds_threshold() {
  return(absolute_diff() >= EULER_TRANSITION_TOLERANCE);
}
